pika/adapters/twisted_connection.py

```diff
@@ -397,416 +337,125 @@ def _adapter_get_write_buffer_size(self):
-        self._reactor.run(installSignalHandlers=False)
-
-    def stop(self):
-        """Implement :py:meth:`.utils.nbio_interface.AbstractIOServices.stop()`.
+        self._on_data_available(data)
 
-        """
-        self._reactor.stop()
 
-    def add_callback_threadsafe(self, callback):
-        """Implement
-        :py:meth:`.utils.nbio_interface.AbstractIOServices.add_callback_threadsafe()`.
+class TwistedProtocolConnection(protocol.Protocol):
+    """A Pika-specific implementation of a Twisted Protocol. Allows using
+    Twisted's non-blocking connectTCP/connectSSL methods for connecting to the
+    server.
 
-        """
-        check_callback_arg(callback, 'callback')
-        self._reactor.callFromThread(callback)
+    It has one caveat: TwistedProtocolConnection objects have a ready
+    instance variable that's a Deferred which fires when the connection is
+    ready to be used (the initial AMQP handshaking has been done). You *have*
+    to wait for this Deferred to fire before requesting a channel.
 
-    def call_later(self, delay, callback):
-        """Implement
-        :py:meth:`.utils.nbio_interface.AbstractIOServices.call_later()`.
+    Since it's Twisted handling connection establishing it does not accept
+    connect callbacks, you have to implement that within Twisted. Also remember
+    that the host, port and ssl values of the connection parameters are ignored
+    because, yet again, it's Twisted who manages the connection.
 
-        """
-        check_callback_arg(callback, 'callback')
-        return _TimerHandle(self._reactor.callLater(delay, callback))
+    """
+    def __init__(self,
+                 parameters=None,
+                 on_close_callback=None):
 
-    def getaddrinfo(self, host, port, on_done, family=0, socktype=0, proto=0,
-                    flags=0):
-        """Implement
-        :py:meth:`.utils.nbio_interface.AbstractIOServices.getaddrinfo()`.
+        self.ready = defer.Deferred()
+        self._impl = _TwistedConnectionAdapter(
+            parameters=parameters,
+            on_open_callback=self.connectionReady,
+            on_open_error_callback=self.connectionFailed,
+            on_close_callback=on_close_callback,
+        )
 
-        """
-        # Use thread pool to run getaddrinfo asynchronously
-        return _TwistedDeferredIOReference(
-            twisted_threads.deferToThreadPool(
-                self._reactor, self._reactor.getThreadPool(),
-                socket.getaddrinfo,
-                # NOTE: python 2.x getaddrinfo only takes positional args
-                host,
-                port,
-                family,
-                socktype,
-                proto,
-                flags),
-            on_done)
-
-    def set_reader(self, fd, on_readable):
-        """Implement
-        :py:meth:`.utils.nbio_interface.AbstractFileDescriptorServices.set_reader()`.
+    def channel(self, channel_number=None):  # pylint: disable-msg=W0221
```

Comment:
@lukebakken, this has been bugging me for some time in Pika core - it provides a seemingly unnecessary arg channel_number that lets the user specify his own channel_number instead of letting pika do it automagically. Is there a real use case for this? Do any of the other clients in RabbitMQ echo-system let you do that?